contracts/governance/Vesting/VestingLogic.sol

```diff
@@ -110,70 +117,73 @@ contract VestingLogic is IVesting, VestingStorage, ApprovalReceiver {
-        require(msg.sender == address(staking), "unauthorized");
-
-        _withdrawTokens(receiver, true);
+    function withdrawTokens(address receiver) public onlyOwners {
+        uint256 startFrom = startDate + cliff;
+        _withdrawTokens(receiver, startFrom, block.timestamp);
     }
 
     /**
-     * @notice Withdraws unlocked tokens from the staking contract and
+     * @notice Withdraws unlocked tokens partially (based on the max withdraw iteration that has been set) from the staking contract and
      * forwards them to an address specified by the token owner.
      * @param receiver The receiving address.
+     * @param startFrom The start value for the iterations.
+     * @param maxWithdrawIterations max withdrawal iteration to work around block gas limit issue.
      * */
-    function withdrawTokens(address receiver) public onlyOwners {
-        _withdrawTokens(receiver, false);
+    function withdrawTokensStartingFrom(
+        address receiver,
+        uint256 startFrom,
+        uint256 maxWithdrawIterations
+    ) public onlyOwners {
+        uint256 defaultStartFrom = startDate + cliff;
+
+        startFrom = _timestampToLockDate(startFrom);
+        startFrom = startFrom < defaultStartFrom ? defaultStartFrom : startFrom;
+
+        // @dev max iterations need to be decreased by 1, otherwise the iteration will always be surplus by 1
+        uint256 maxWithdrawDate = (startFrom + (FOUR_WEEKS * (maxWithdrawIterations.sub(1))));
+        uint256 endAt = endDate < maxWithdrawDate ? endDate : maxWithdrawDate;
+        _withdrawTokens(receiver, startFrom, endAt);
     }
 
     /**
      * @notice Withdraws tokens from the staking contract and forwards them
      * to an address specified by the token owner. Low level function.
      * @dev Once here the caller permission is taken for granted.
      * @param receiver The receiving address.
-     * @param isGovernance Whether all tokens (true)
+     * @param startFrom start withdrawal from date.
+     * @param endAt end time for regular withdrawal
      * or just unlocked tokens (false).
      * */
-    function _withdrawTokens(address receiver, bool isGovernance) internal {
+    function _withdrawTokens(
+        address receiver,
+        uint256 startFrom,
+        uint256 endAt
+    ) internal {
         require(receiver != address(0), "receiver address invalid");
 
         uint96 stake;
 
-        /// @dev Usually we just need to iterate over the possible dates until now.
-        uint256 end;
-
-        /// @dev In the unlikely case that all tokens have been unlocked early,
-        ///   allow to withdraw all of them.
-        if (staking.allUnlocked() || isGovernance) {
-            end = endDate;
-        } else {
-            end = block.timestamp;
-        }
+        uint256 end = endAt < block.timestamp ? endAt : block.timestamp;
+        if (end > endDate && block.timestamp > endDate) end = endDate;
 
```

Comment:
why if(staking.allUnlocked() is removed? it should be able to process till the end, not till block.timestamp as mentioned in this comment.

Reply:
I thought the initial condition was `if(staking.allUnlocked && isGovernance)` that's why i removed, apparently it OR condition.
Revert back the changes in this [commit](https://github.com/DistributedCollective/Sovryn-smart-contracts/pull/529/commits/eaf61228f6677387d6ee5b37b08ed83a61ceb269)